### PR TITLE
✨(dashboard) add video event distribution during the video

### PR DESCRIPTION
**Pull request by SergioSim**
## Purpose

The video event distribution during the video could be a useful indicator in the video details dashboard.

## Proposal

This panel splits the video timeline in equal intervals and shows the event distribution by the type of interaction.

### CODE REVIEW

**Comment by SergioSim** [comment]
There are two graphs to choose from:
Screenshot from 2021-06-25 08-04-38
The first is much simpler (and faster) but lacks colors and don't show clearly the groups.
The second makes 6 queries just to render the colors and groups but the x-axis label is incorrect. 
Both panels don't show the video time on the x-axis (but it appears in the tooltip on hover):
Screenshot from 2021-06-25 08-05-23

**Comment by SergioSim** [comment]
Just in case, have added the "plan B" option with 5 separate panels:
Screenshot from 2021-06-25 12-56-40
I continue to search for a stacked option.

**Comment by SergioSim** [comment]
Now the dashboard is using the barplot plugin (as a submodule).
Screenshot from 2021-06-28 17-21-41

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "etc/grafana/provisioning/plugins/groupedbarchart-panel"]
+	path = etc/grafana/provisioning/plugins/groupedbarchart-panel
+	url = https://github.com/gipong/grafana-groupedbarchart-panel.git
```
Comment (jmaupetit) [inline .gitmodules:3]: Git submodule is evil :smiling_imp: I'd prefer to `curl` the project's master state [1] in a `plugins` Makefile rule WDYT?

[1] https://github.com/gipong/grafana-groupedbarchart-panel/archive/refs/heads/master.zip

File: src/dashboards/videos/details.jsonnet [hunk 2]
```diff
@@ -264,3 +274,120 @@ dashboard.new(
...
+    aliasColors: {},
+    avgLineShow: false,
+    barValuesShow: true,
+    cacheTimeout: null,
+    chartType: 'stacked bar chart',
+    colorSch: [],
+    colorSet: [
+      {
+        '$$hashKey': 'object:204',
+        color: '#1f77b4',
```
Comment (jmaupetit) [inline src/dashboards/videos/details.jsonnet:294]: Chosen colors belongs to the default grafana 8 palette?

**Review by jmaupetit — COMMENTED** [review]
Looks great :muscle: 

Don't forget the changelog :wink:

File: src/dashboards/videos/details.jsonnet (hunk 2 above)
Comment (quitterie-lcs) [reply src/dashboards/videos/details.jsonnet:294]: I think that as it is directly a JSON the color has to be specified

File: .gitmodules (hunk 1 above)
Comment (SergioSim) [reply .gitmodules]: Great Idea) Thanks) Updated the Makefile)

File: src/dashboards/videos/details.jsonnet (hunk 2 above)
Comment (SergioSim) [reply src/dashboards/videos/details.jsonnet]: I tried to replace the colors with the default grafana palette but for some reason the plugin seems to ignore this configuration. I thought too that it has to be specified, but surprisingly, removing this part didn't change the output, so removed it.

File: Makefile [hunk 3]
```diff
@@ -64,6 +68,13 @@ logs: ## display grafana logs (follow mode)
 	@$(COMPOSE) logs -f grafana
 .PHONY: logs
 
+plugins: ## download plugins
+	mkdir -p ./var/lib/grafana/plugins
```
Comment (jmaupetit) [inline Makefile:72]: I would add a `var/lib/grafana/plugins` rule in the FILES section, add this rule to the tree rule dependencies and make the `plugins` rule depend on the `tree` rule.

File: Makefile [hunk 4]
```diff
@@ -64,6 +68,13 @@ logs: ## display grafana logs (follow mode)
...
+	curl -Lo /tmp/grafana_bar_chart_plugin.zip $(BAR_CHART_PLUGIN)
+	rm -rf ./var/lib/grafana/plugins/*
+	unzip /tmp/grafana_bar_chart_plugin.zip -d ./var/lib/grafana/plugins/
```
Comment (jmaupetit) [inline Makefile:75]: We can make a rule per plugin and make the `plugins` rule a meta phony rule that installs them all. WDYT?

File: Makefile (hunk 3 above)
Comment (SergioSim) [reply Makefile]: Good Idea) Thanks) Updated this part) the `var/lib/grafana/plugins` rule should just create the plugins directory ?

Comment (jmaupetit) [reply Makefile]: Yes.

File: CHANGELOG.md [hunk 5]
```diff
@@ -10,6 +10,7 @@ Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to
 
 ### Added
 
+- Video event distribution during the video panel in details Dashboard
```
Comment (jmaupetit) [inline CHANGELOG.md:13]:
```suggestion
- Video details dashboard new panel: video events distribution along the video timeline
```

File: Makefile (hunk 4 above)
Comment (SergioSim) [reply Makefile]: Good Idea) Thanks)  Updated)

File: Makefile [hunk 6]
```diff
@@ -25,8 +28,17 @@ var/lib/grafana/%.json: src/%.jsonnet
 	mkdir -p $(shell dirname $@)
 	bin/jsonnet -o /$@ $<
 
+var/lib/grafana/plugins:
+	mkdir -p ./var/lib/grafana/plugins
+
+var/lib/grafana/plugins/grafana-groupedbarchart-panel-master:
+	curl -Lo /tmp/grafana_bar_chart_plugin.zip $(BAR_CHART_PLUGIN)
+	rm -rf ./var/lib/grafana/plugins/*
+	unzip /tmp/grafana_bar_chart_plugin.zip -d ./var/lib/grafana/plugins/
```
Comment (jmaupetit) [inline Makefile:37]: Maybe we can put this in a plugins section

File: Makefile [hunk 7]
```diff
@@ -25,8 +28,17 @@ var/lib/grafana/%.json: src/%.jsonnet
...
+
 tree: \
-	var/lib/grafana
+	var/lib/grafana \
+	plugins
```
Comment (jmaupetit) [inline Makefile:41]:
```suggestion
	var/lib/grafana \
	var/lib/grafana/plugins
```

File: Makefile [hunk 8]
```diff
@@ -64,6 +76,12 @@ logs: ## display grafana logs (follow mode)
 	@$(COMPOSE) logs -f grafana
 .PHONY: logs
 
+plugins: \
+	var/lib/grafana/plugins \
```
Comment (jmaupetit) [inline Makefile:80]:
```suggestion
	tree \
```

File: CHANGELOG.md (hunk 5 above)
Comment (SergioSim) [reply CHANGELOG.md:13]: Thanks)

File: Makefile [hunk 9]
```diff
@@ -25,8 +28,17 @@ var/lib/grafana/%.json: src/%.jsonnet
 	mkdir -p $(shell dirname $@)
 	bin/jsonnet -o /$@ $<
 
+var/lib/grafana/plugins:
+	mkdir -p ./var/lib/grafana/plugins
+
+var/lib/grafana/plugins/grafana-groupedbarchart-panel-master:
+	curl -Lo /tmp/grafana_bar_chart_plugin.zip $(BAR_CHART_PLUGIN)
+	rm -rf ./var/lib/grafana/plugins/*
```
Comment (jmaupetit) [inline Makefile]: Maybe we should add this to the `clean` rule, WDYT?

File: Makefile (hunk 6 above)
Comment (SergioSim) [reply Makefile:45]: Yes) Thanks) Added the section)

File: Makefile (hunk 9 above)
Comment (SergioSim) [reply Makefile]: Nice catch) Thanks) Indeed, this should not be here)

File: Makefile [hunk 10]
```diff
@@ -58,12 +57,18 @@ build: ## build potsie development image
...
+.PHONY: clean
+
 compile: \
 	tree \
 	$(targets)
 compile: ## compile jsonnet sources to json
 .PHONY: compile
 
+down: \
+	clean
```
Comment (jmaupetit) [inline Makefile:71]: I would have done the opposite (`down` as a `clean` dependency) as `clean` is a more popular / expected rule for a fresh start. WDYT?

File: Makefile (hunk 9 above)
Comment (SergioSim) [reply Makefile]: Added the `clean` rule which removes the `var` directory)

File: Makefile [hunk 11]
```diff
@@ -67,10 +69,9 @@ compile: \
 compile: ## compile jsonnet sources to json
 .PHONY: compile
 
-down: \
-	clean
 down: ## remove stack (warning: it removes the database container)
-	@$(COMPOSE) down
+	# FIXME: docker-compose down fails because of potsie network attached to elastic search
+	@$(COMPOSE) down || echo WARNING: unable to remove the stack
```
Comment (jmaupetit) [inline Makefile:74]: Is it a network related issue? AFAIR I did not encounter this issue.

File: Makefile (hunk 10 above)
Comment (SergioSim) [reply Makefile]: Sounds good) Thanks) updated this part)

File: Makefile (hunk 11 above)
Comment (SergioSim) [reply Makefile:74]: In my case this happens because `elasticsearch` depends on the `potsie` network and `docker-compose down` tries to remove it.

Comment (SergioSim) [reply Makefile:74]: A solution is to manually stop ralphs `elasticsearch` container and then run `make down` on potsie. Will update the comment)

File: Makefile [hunk 12]
```diff
@@ -70,8 +70,7 @@ compile: ## compile jsonnet sources to json
 .PHONY: compile
 
 down: ## remove stack (warning: it removes the database container)
-	# FIXME: docker-compose down fails because of potsie network attached to elastic search
-	@$(COMPOSE) down || echo WARNING: unable to remove the stack
+	@$(COMPOSE) down || echo WARNING: unable to remove the stack. Try to stop ralph's elasticsearch container
```
Comment (jmaupetit) [inline Makefile:73]:
```suggestion
	@$(COMPOSE) down || echo WARNING: unable to remove the stack. Try to stop linked containers or networks first.
```

Comment (SergioSim) [reply Makefile:73]: Thanks)

File: src/dashboards/videos/details.jsonnet [hunk 13]
```diff
@@ -449,3 +458,77 @@ dashboard.new(
...
+      The interval size is controlled by the `Event group interval` variable
+    |||,
+    aliasColors: {},
+    avgLineShow: false,
+    barValuesShow: true,
+    cacheTimeout: null,
+    chartType: 'stacked bar chart',
+    colorSch: [],
+    datasource: 'lrs',
+    fontColor: '#fff',
```
Comment (SergioSim) [inline src/dashboards/videos/details.jsonnet:476]: Apropos, we can control the font color, do we set it to `#000` for the light grafana theme ?

Comment (jmaupetit) [reply src/dashboards/videos/details.jsonnet:476]: Nope. Leave it as is for now. You'll do better in your own plugin using this: https://grafana.com/docs/grafana/latest/packages_api/ui/withtheme/

:wink:

**Review by jmaupetit — APPROVED** [review]

**Pull request merged**